tsweb/server.go

```diff
@@ -0,0 +1,345 @@
+// Copyright (c) 2022 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package tsweb
+
+import (
+	"context"
+	"crypto/tls"
+	"expvar"
+	"fmt"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"golang.org/x/crypto/acme/autocert"
+	"tailscale.com/metrics"
+	"tailscale.com/types/logger"
+)
+
+const defaultTimeout = 10 * time.Second
+
+// ServerConfig is the initial configuration of a Server, for
+// consumption by NewServer.
+type ServerConfig struct {
+	// Name is a human-readable, alphanumeric name for the HTTP
+	// server.
+	// The name is for internal use only, for things like naming cache
+	// directories, identifying the source of autocert emails,
+	// human-readable debug pages, ...
+	Name string
+	// Addr specifies the TCP address for the server to listen on, in
+	// the form "host:port". If Addr specifies port 443, TLS serving
+	// is automatically configured and a second listener is set up on
+	// port 80 to handle HTTP to HTTPS redirection. All other ports
+	// result in HTTP-only serving.
+	Addr string
+	// Handler is the HTTP.Handler to invoke to handle
+	// requests. Cannot be nil.
+	Handler http.Handler
+	// Hostnames is the list of hostnames that the server accepts for
+	// serving TLS requests.
+	Hostnames []string
+
+	// HSTSNoSubdomains is whether to tell browsers to only apply
+	// strict HTTPS serving to the current domain being requested,
+	// rather than the request domain and all its subdomains (the
+	// default).
+	HSTSNoSubdomains bool
+
+	// RequestLog is a log sink to write access logs for requests
```

Comment:
this loses all the structure, though?  (byte counts, http status type, etc)

Reply:
Switched to a structured callback. Again my goal here was that everything currently writes to a Logf, so by having a structured callback here I'm forcing the same boilerplate onto everyone in order to logf it. Maybe that won't be terrible, I dunno.